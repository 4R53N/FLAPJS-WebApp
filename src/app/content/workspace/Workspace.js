import React from 'react';
import './Workspace.css';

import Config from 'config.js';

import Subtitle from './Subtitle.js';

import NodalGraphRenderer from 'modules/fsa/graph/renderer/NodalGraphRenderer.js';
import SelectionBoxRenderer from './renderer/SelectionBoxRenderer.js';
import HighlightRenderer from './renderer/HighlightRenderer.js';

const WORKSPACE_OFFSET_X = 0;
const WORKSPACE_OFFSET_Y = 0;
const EXPORT_PADDING_X = 30;
const EXPORT_PADDING_Y = 0;

class Workspace extends React.Component
{
  constructor(props)
  {
    super(props);

    this.ref = React.createRef();
  }

  getSVGForExport(width, height)
  {
    const svg = this.ref;
    const viewport = this.props.inputController.getViewport();
    const offsetX = viewport.getOffsetX();
    const offsetY = viewport.getOffsetY();
    const bounds = this.props.graphController.getGraph().getBoundingRect();
    const dx = bounds.minX + offsetX - EXPORT_PADDING_X;
    const dy = bounds.minY + offsetY - EXPORT_PADDING_Y;
    const w = bounds.width + EXPORT_PADDING_X * 2;
    const h = bounds.height + EXPORT_PADDING_Y * 2;
    const clone = svg.cloneNode(true);
    clone.setAttribute('viewBox', dx + " " + dy + " " + w + " " + h);
    clone.setAttribute('width', width);
    clone.setAttribute('height', height);

    //Apply the workspace font (refer to Workspace.css)
    clone.setAttribute('font-size', "16px");
    clone.setAttribute('font-family', "monospace");

    /*
    //TODO: Link the font family to svg
    const link = document.createElement("link");
    link.setAttribute("rel", "stylesheet");
    clone.appendChild(link);
    */

    //Remove unwanted ui elements from image
    const uiElements = clone.getElementsByClassName("graph-ui");
    for(const e of uiElements)
    {
      e.remove();
    }

    return clone;
  }

  render()
  {
    const graphController = this.props.graphController;
    const inputController = this.props.inputController;
    const machineController = this.props.machineController;
    const tester = this.props.tester;

    const graph = graphController.getGraph();
    const viewport = inputController.getViewport();
    const machineBuilder = machineController.getMachineBuilder();
    const picker = inputController.getPicker();
    const selectionBox = picker.getSelectionBox();

    let size = Config.DEFAULT_GRAPH_SIZE * Math.max(Number.MIN_VALUE, viewport.getScale());
    const halfSize = size / 2;

    //Must not be a block content (must inline)
    return <svg id="workspace-content" ref={ref=>this.ref=ref}
      viewBox={(-halfSize + WORKSPACE_OFFSET_X) + " " + (-halfSize + WORKSPACE_OFFSET_Y) + " " + size + " " + size}
      xmlns="http://www.w3.org/2000/svg">

      {/* Graph subtitle */}
      <Subtitle visible={graph.isEmpty()}/>

      {/* Graph elements */}
      <g id="workspace-content-elements" transform={
        "translate(" +
        viewport.getOffsetX() + " " +
        viewport.getOffsetY() + ")"}>

        {/* Graph origin crosshair */}
        <line className="graph-ui" x1="0" y1="-5" x2="0" y2="5" stroke="rgba(0,0,0,0.04)"/>
        <line className="graph-ui" x1="-5" y1="0" x2="5" y2="0" stroke="rgba(0,0,0,0.04)"/>

        {/* Graph objects */}
        <NodalGraphRenderer graph={graph} inputController={inputController}/>

        {/* Graph GUIs */}
        <g>
          {/* Selected elements */}
          { picker.hasSelection() &&
            picker.getSelection(graph).map((e, i) =>
              <HighlightRenderer key={e.getGraphElementID()} className={inputController.isTrashMode() ? "highlight-error" : "highlight-select"} target={e} type="node"/>) }

          {/* Selection box */}
          <SelectionBoxRenderer visible={selectionBox.visible}
            fromX={selectionBox.fromX} fromY={selectionBox.fromY}
            toX={selectionBox.toX} toY={selectionBox.toY}/>

          {/* Node warning targets */}
          { machineController.getMachineBuilder().machineErrorChecker.warningNodes.map((e, i) =>
            <HighlightRenderer key={e.getGraphElementID()} className="highlight-warning graph-gui" target={e} type="node" offset="6"/>) }

          {/* Edge warning targets */}
          { machineController.getMachineBuilder().machineErrorChecker.warningEdges.map((e, i) =>
            <HighlightRenderer key={e.getGraphElementID()} className="highlight-warning graph-gui" target={e} type="edge" offset="6"/>) }

          {/* Node error targets */}
          { machineController.getMachineBuilder().machineErrorChecker.errorNodes.map((e, i) =>
            <HighlightRenderer key={e.getGraphElementID()} className="highlight-error graph-gui" target={e} type="node" offset="6"/>) }

          {/* Edge error targets */}
          { machineController.getMachineBuilder().machineErrorChecker.errorEdges.map((e, i) =>
            <HighlightRenderer key={e.getGraphElementID()} className="highlight-error graph-gui" target={e} type="edge" offset="6"/>) }


          {/* Node test targets */}
          { tester.testMode.targets.map((e, i) => {
              return <HighlightRenderer key={e.getGraphElementID()} className="highlight-test graph-gui" target={e} type="node" offset="6"/>;
            }) }

          {/* Hover markers */}
          { picker.hasTarget() &&
            !picker.isTargetInSelection() &&
            <HighlightRenderer className={inputController.isTrashMode() ? "highlight-error" : "highlight-select"} target={picker.target} type={picker.targetType}/> }

        </g>
      </g>
    </svg>;
  }
}

export default Workspace;
