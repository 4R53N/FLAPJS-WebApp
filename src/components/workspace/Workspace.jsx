import React from 'react';
import PropTypes from 'prop-types';
import Style from './Workspace.module.css';

/**
 * A React component that can do anything you want. :D
 */
class Workspace extends React.Component
{
    constructor(props)
    {
        super(props);

        this.state = {
            open: true,
            full: false,
            tabIndex: 0,
        };
        
        this.container = React.createRef();
        this.drawer = React.createRef();

        this.prevDrawerHandlePosition = { x: 0, y: 0 };
        this.onDrawerHandleDragBegin = this.onDrawerHandleDragBegin.bind(this);
        this.onDrawerHandleDragMove = this.onDrawerHandleDragMove.bind(this);
        this.onDrawerHandleDragEnd = this.onDrawerHandleDragEnd.bind(this);
    }

    onDrawerHandleDragBegin(e)
    {
        const drawerSize = this.drawer.current.getBoundingClientRect();
        this.prevDrawerHandlePosition.x = e.clientX + drawerSize.width;
        this.prevDrawerHandlePosition.y = e.clientY + drawerSize.height;

        document.addEventListener('mousemove', this.onDrawerHandleDragMove);
        document.addEventListener('mouseup', this.onDrawerHandleDragEnd);
    }

    onDrawerHandleDragMove(e)
    {
        const dx = this.prevDrawerHandlePosition.x - e.clientX;
        this.workspace.current.style.setProperty('--drawer-width', `${dx}px`);
    }

    onDrawerHandleDragEnd(e)
    {
        document.removeEventListener('mousemove', this.onDrawerHandleDragMove);
        document.removeEventListener('mouseup', this.onDrawerHandleDragEnd);
    }

    openDrawer(tabIndex = 0)
    {
        if (!this.state.open)
        {
            this.setState({ open: true, tabIndex });
        }
    }

    closeDrawer()
    {
        if (this.state.open)
        {
            this.setState({ open: false });
        }
    }

    toggleDrawer(tabIndex = 0)
    {
        this.setState(prev =>
        {
            return { open: !prev.open, tabIndex };
        });
    }

    /** @override */
    render()
    {
        const props = this.props;
        const open = this.state.open;

        const renderSideBar = props.renderSideBar;
        const renderViewport = props.renderViewport;
        const renderDrawer = props.renderDrawer;
        const renderForeground = props.renderForeground;
        const renderBackground = props.renderBackground;
        
        return (
            <section ref={this.container}
                className={Style.container + (props.className || '')}>
                <div className={Style.foreground}>
                    <nav className={Style.sidebar
                        + ' primary '}>
                        {renderSideBar && renderSideBar(this)}
                    </nav>
                    <div className={Style.viewport}>
                        {renderViewport && renderViewport(this)}
                    </div>
                    <div ref={this.drawer} className={Style.drawer + (open ? ' open ' : '')}>
                        <div className={Style.handle}
                            onMouseDown={this.onDrawerHandleDragBegin}
                            role="presentation">
                        </div>
                        <div className={Style.content}>
                            {renderDrawer && renderDrawer(this)}
                        </div>
                    </div>
                    <nav className={Style.sidebar
                        + ' secondary '}>
                        {renderSideBar && renderSideBar(this)}
                    </nav>
                    {renderForeground && renderForeground(this)}
                </div>
                <div className={Style.background}>
                    {renderBackground && renderBackground(this)}
                </div>
            </section>
        );
    }
}

Workspace.propTypes = {
    className: PropTypes.string,
    renderForeground: PropTypes.func,
    renderBackground: PropTypes.func,
    renderViewport: PropTypes.func,
    renderDrawer: PropTypes.func,
    renderSideBar: PropTypes.func,
};
Workspace.defaultProps = {
};

export default Workspace;
