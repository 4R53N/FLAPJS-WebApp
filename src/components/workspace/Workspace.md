# Workspace

## Motivation
Cause I love all of you. <3

## Work Log
- Created Workspace component.