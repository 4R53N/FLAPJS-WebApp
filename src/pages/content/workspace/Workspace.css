#workspace-content {
  padding: 0;
  margin: 0;

  background: none;
  /*background: var(--color-workspace-main);*/
}
.graph-node {
  fill: var(--color-graph-node);
  stroke: var(--color-workspace-text);
}
.graph-node-inner {
  stroke: var(--color-workspace-text);
}
.graph-node-label {
  text-anchor: middle;
  font-size: inherit;
  font-family: 'Overpass Mono', monospace;
}
.graph-edge {
  stroke: black;
}
.graph-edge-label {
  text-anchor: middle;
  font-size: inherit;
  font-family: 'Overpass Mono', monospace;
}
.graph-initial {
  stroke: var(--color-workspace-text);
}
.graph-selection {
  fill-opacity: 0.1;
  fill: var(--color-selection-main);
  stroke: var(--color-selection-border);
}
.highlight-select, .highlight-test {
  stroke: var(--color-highlight-select);
}
.highlight-error {
  stroke: var(--color-viewport-error);
}
