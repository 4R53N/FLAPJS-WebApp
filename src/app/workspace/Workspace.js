import React from 'react';

import './Workspace.css';

import * as Config from 'config.js';

import NodeRenderer from './renderer/NodeRenderer.js';
import EdgeRenderer from './renderer/EdgeRenderer.js';
import SelectionBoxRenderer from './renderer/SelectionBoxRenderer.js';
import InitialMarkerRenderer from './renderer/InitialMarkerRenderer.js';
import BorderRenderer from './renderer/BorderRenderer.js';

class Workspace extends React.Component
{
  constructor(props)
  {
    super(props);
    this.viewpowrt = React.createRef();
  }

  componentDidMount()
  {
    //Initialize the controller to graph components
    this.props.controller.initialize(this.viewport);
  }

  componentDidUpdate()
  {
    //Update input controller (usually mouse position for hover info)
    this.props.controller.onUpdate();
  }

  render()
  {
    const graph = this.props.graph;
    const controller = this.props.controller;

    //Must not be a block content (must inline)
    return <svg id="workspace-content" ref={ref => this.viewport = ref}
      viewBox="-150 -150 300 300"
      xmlns="http://www.w3.org/2000/svg">
      //Graph objects
      <g>
        //Nodes
        { graph.nodes.map((e, i) =>
          <NodeRenderer key={i} node={e}/>) }

        //Edges
        { graph.edges.map((e, i) =>
          <EdgeRenderer key={i} edge={e}
            start={e.getStartPoint()}
            end={e.getEndPoint()}
            center={e.getCenterPoint()}
            label={e.label}/>) }
      </g>

      //Graph guis
      { controller != null &&
        <g>
          //Initial marker (and ghost)
          { graph.getStartNode() && (controller.ghostInitialMarker == null ?
            <InitialMarkerRenderer node={graph.getStartNode()}/> :
            <InitialMarkerRenderer node={controller.ghostInitialMarker}/>) }

          //Selected Elements
          { controller.selector.hasSelection() &&
            controller.selector.getSelection().map((e, i) =>
              <Select key={i} target={e} type="node"/>) }

          //SelectionBox
          <SelectionBoxRenderer src={controller.selector}/>

          //Hover Element
          { controller.pointer.target &&
            !controller.selector.isTargetSelected(controller.pointer.target) &&
            <Select target={controller.pointer.target} type={controller.pointer.targetType}/> }

          //Border Element
          <BorderRenderer mode={
            controller.pointer.isTrashMode() ? 2 :
            controller.pointer.isMoveMode() ? 1 : 0}/>

        </g> }
    </svg>;
  }
}

function Select(props)
{
  const target = props.target;
  const type = props.type;

  let x = 0;
  let y = 0;
  let r = Config.CURSOR_RADIUS;
  switch(type)
  {
    case "node":
      x = target.x;
      y = target.y;
      r = Config.NODE_RADIUS;
      break;
    case "edge":
      const center = target.getCenterPoint();
      x = center.x;
      y = center.y;
      r = Config.EDGE_RADIUS;
      break;
    case "endpoint":
      const endpoint = target.getEndPoint();
      x = endpoint.x;
      y = endpoint.y;
      r = Config.ENDPOINT_RADIUS;
      break;
    case "initial":
      x = target.x - Config.NODE_RADIUS;
      y = target.y;
      r = Config.CURSOR_RADIUS;
      break;
  }

  return <g>
    <circle
      cx={x}
      cy={y}
      r={r + Config.HOVER_RADIUS_OFFSET}
      strokeDasharray={Config.HOVER_LINE_DASH}
      strokeWidth={Config.HOVER_LINE_WIDTH}
      stroke={Config.HOVER_STROKE_STYLE}
      fill="none" />
  </g>;

  /*
    const angle = hoverAngle;
    ctx.strokeStyle = Config.HOVER_STROKE_STYLE;
    ctx.lineWidth = Config.HOVER_LINE_WIDTH;
    ctx.beginPath();
    ctx.setLineDash(Config.HOVER_LINE_DASH);
    ctx.arc(x, y, r + Config.HOVER_RADIUS_OFFSET, 0 + angle, Config.PI2 + angle);
    ctx.stroke();
  */
}

export default Workspace;
