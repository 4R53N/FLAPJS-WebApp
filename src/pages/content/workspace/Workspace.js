import React from 'react';
import './Workspace.css';

import Config from 'config.js';

import Subtitle from './Subtitle.js';

import NodeRenderer from './renderer/NodeRenderer.js';
import EdgeRenderer from './renderer/EdgeRenderer.js';
import SelectionBoxRenderer from './renderer/SelectionBoxRenderer.js';
import InitialMarkerRenderer from './renderer/InitialMarkerRenderer.js';
import HighlightRenderer from './renderer/HighlightRenderer.js';

const WORKSPACE_OFFSET_X = 15;
const WORKSPACE_OFFSET_Y = 0;

class Workspace extends React.Component
{
  constructor(props)
  {
    super(props);

    this.ref = React.createRef();
  }

  componentWillUpdate()
  {
    //Update input controller (usually mouse position for hover info)
    //this.props.controller.pointer.updateTarget();
  }

  render()
  {
    const graph = this.props.graph;
    const controller = this.props.controller;
    const pointer = controller.pointer;
    const machineBuilder = this.props.machineBuilder;

    let size = Config.DEFAULT_GRAPH_SIZE * Math.max(Number.MIN_VALUE, pointer.scale);
    const halfSize = size / 2;

    //Must not be a block content (must inline)
    return <svg id="workspace-content" ref={ref=>this.ref=ref}
      viewBox={(-halfSize + WORKSPACE_OFFSET_X) + " " + (-halfSize + WORKSPACE_OFFSET_Y) + " " + size + " " + size}
      xmlns="http://www.w3.org/2000/svg">

      <Subtitle visible={graph.isEmpty()}/>

      <filter id="error-highlight" height="300%" width="300%" x="-75%" y="-75%">
        <feColorMatrix type="matrix"
          result="color"
          values={
            "1 0 0 0 1 " +
            "0 0 0 0 0 " +
            "0 0 0 0 0 " +
            "0 0 0 1 0"
          }/>
        <feGaussianBlur in="color" stdDeviation="1" result="blur"/>
        <feMerge>
          <feMergeNode in="blur"/>
          <feMergeNode in="SourceGraphic"/>
        </feMerge>
      </filter>

      <g transform={"translate(" + controller.pointer.offsetX + " " + controller.pointer.offsetY + ")"}>

        <line x1="0" y1="-5" x2="0" y2="5" stroke="rgba(0,0,0,0.04)"/>
        <line x1="-5" y1="0" x2="5" y2="0" stroke="rgba(0,0,0,0.04)"/>

        //Graph objects
        <g>
          //Nodes
          { graph.nodes.map((e, i) =>
            <NodeRenderer key={i} node={e}/>) }

          //Edges
          { graph.edges.map((e, i) =>
            <EdgeRenderer key={i} edge={e}
              start={e.getStartPoint()}
              end={e.getEndPoint()}
              center={e.getCenterPoint()}
              label={e.label}/>) }
        </g>

        //Graph guis
        <g>
          //Initial marker (and ghost)
          { graph.getStartNode() && (controller.ghostInitialMarker == null ?
            <InitialMarkerRenderer node={graph.getStartNode()}/> :
            <InitialMarkerRenderer node={controller.ghostInitialMarker}/>) }

          //Selected Elements
          { controller.selector.hasSelection() &&
            controller.selector.getSelection().map((e, i) =>
              <HighlightRenderer key={e.label} target={e} type="node" color="gray"/>) }

          //SelectionBox
          <SelectionBoxRenderer src={controller.selector}/>

          { machineBuilder.machineErrorChecker.errorNodes.map((e, i) =>
            <HighlightRenderer key={e.label} target={e} type="node" color="red" offset="6"/>) }

          //Hover Element
          { /*controller.pointer.target &&
            !controller.selector.isTargetSelected(controller.pointer.target) &&
            <Select target={controller.pointer.target} type={controller.pointer.targetType}/>*/ }

          </g>
        </g>
    </svg>;
  }
}

export default Workspace;
