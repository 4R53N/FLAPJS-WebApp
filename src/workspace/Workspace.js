import React from 'react';

import './Workspace.css';

class Workspace extends React.Component
{
  constructor(props)
  {
    super(props);
  }

  render()
  {
    return <svg id="workspace-content"
      viewBox="0 0 100 10"
      xmlns="http://www.w3.org/2000/svg">
      <circle cx="50" cy="5" r="10"/>
    </svg>;
  }
}

export default Workspace;
