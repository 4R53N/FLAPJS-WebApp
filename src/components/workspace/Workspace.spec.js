import React from 'react';
import { shallow } from 'enzyme';

import Workspace from './Workspace.jsx';

let wrapper;

beforeEach(() =>
{
    wrapper = shallow(<Workspace/>);
});

describe('<Workspace/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<Workspace/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<Workspace/> lifecycle', () =>
{
    /** ...later. */
});
