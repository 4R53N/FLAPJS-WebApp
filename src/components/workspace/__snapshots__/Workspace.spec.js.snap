// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Workspace/> rendering renders correctly 1`] = `
<section
  className="container"
>
  <div
    className="foreground"
  >
    <nav
      className="sidebar primary "
    />
    <div
      className="viewport"
    />
    <div
      className="drawer open "
    >
      <div
        className="handle"
        onMouseDown={[Function]}
        role="presentation"
      />
      <div
        className="content"
      />
    </div>
    <nav
      className="sidebar secondary "
    />
  </div>
  <div
    className="background"
  />
</section>
`;
